### ISSUE TITLE
Database-like transactions for actions like creating rooms, enabling rollbacks

### ISSUE THREAD

**Issue by stoically** [matrix-org/synapse#13542]
**Description:**

In my use case I'm creating a matrix room and store the returned room id into a database. In case something goes wrong while storing the room id into the database – or somewhere else – I'm left with a dangling matrix room.

If synapse would support database-like transactions I could create the room, store the room id and only after successfully storing it, commit the room creation transaction.

**Comment by DMRobertson** [comment]
I think this is a duplicate of #8895. To [quote](https://github.com/matrix-org/synapse/issues/8895#issuecomment-742564480) rich:

> It boils down to the fact that createRoom is not atomic, and exceptions (for any reason) in the middle of it give you a broken room.

**Comment by richvdh** [comment]
I don't think this is quite a duplicate of #8895 - it sounds more like the reporter wants transactions at the matrix API level (it's not entirely clear tbh).

But:
 (a) That would be a protocol-level change so would be more appropriate in matrix-spec
 (b) I can't see it happening any time soon anyway
 (c) I'm not sure it even helps: what if there is an error committing the room creation transaction? Your database is still out of sync with synapse, it's just the other way around.

**Comment by DMRobertson** [comment]
Oh I see. I missed that OP was talking about an external database, not Synapse's.

**Comment by stoically** [comment]
@richvdh You're right. Sorry for not being clearer.

> (a) That would be a protocol-level change so would be more appropriate in matrix-spec

Right, that makes sense. I'll open an issue in the spec proposals repo accordingly.

> (c) I'm not sure it even helps: what if there is an error committing the room creation transaction? Your database is still out of sync with synapse, it's just the other way around.

For the database I have transactions available. So in my mind, if committing the room creation fails, the database insert transaction will never be committed and rolled back accordingly.

**Comment by stoically** [comment]
> So in my mind, if committing the room creation fails, the database insert transaction will never be committed and rolled back accordingly.

Actually, that probably doesn't work.

Thinking about it more: With Matrix API level transactions I could prevent a dangling matrix room, but might be left with a dangling entry in my database if committing the synapse transaction fails. The advantage would be: I know which role the database entry plays in my domain context and can handle the fact that the room doesn't actually exist properly. While, if I end up with a dangling matrix room I don't have domain context attached to it and can only guess which rooms are dangling and can be cleaned up.

**Comment by richvdh** [comment]
this isn't really the place to discuss further, but a complete solution to your problem probably involves some sort of "transaction ID" so that you can safely make multiple calls to `/createRoom` if the first one fails (in much the same way as you can with individual messages). In that way, you can set up your database entry *first* to make sure that you have the context you need.

But this is getting way into sci-fi. You can open an issue at matrix-spec if you like, but I don't think anyone is going to drive this for you. You'd probably have to prototype it yourself and open an MSC (see https://github.com/matrix-org/matrix-spec-proposals#matrix-specification-proposals).
